Use control schema for messages

FloGa opened the pull request:
According to the specification for [UPnP Devices][1], the arguments for the control actions need to match the order of arguments as desribed in the schema. To achieve that, we retrieve the schemas when searching for a compitable gateway and use the appropriate schema for each action to build the arguments properly.

Fixes #48.
    
[1]: https://openconnectivity.org/upnp-specs/UPnP-arch-DeviceArchitecture-v2.0-20200417.pdf

sbstp commented on src/aio/search.rs:
```diff
@@ -95,81 +107,147 @@ impl SearchFuture {
+        Ok(urls)
+    }
+
+    fn request_schemas(
```
Do you think that using async would make the code simpler and cleaner here? I have no issue whatsoever with dropping the raw future implementations and using async functions for clarity.

FloGa replied:
Yes, definitely. I just wanted to preserve as much of your code as possible, so as not to rewrite the whole story. But if you are okay with it, I will try a more `async` approach, without the `Future` implementation.

sbstp commented on src/aio/search.rs:
```diff
@@ -95,81 +107,147 @@ impl SearchFuture {
 impl Future for SearchFuture {
     type Output = Result<Gateway, SearchError>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<Gateway, SearchError>> {
```
If you'd like to turn this into an async function as well, I think it could be much cleaner. You don't have to if you don't feel like it though.

FloGa replied:
Yes, I can rewrite it. As I said, I wanted to preserve much of the code, more so because it was my first time actually working with asynchronous Rust and I didn't know the best practices. But I am positive that I can make it more readable with a nice `async/.await` approach.

sbstp commented on src/common/messages.rs:
```diff
@@ -35,90 +46,103 @@ pub fn format_get_external_ip_message() -> String {
+                "NewPortMappingDescription" => description.to_string(),
+                "NewProtocol" => protocol.to_string(),
+                "NewRemoteHost" => "".to_string(),
+                _ => panic!("Nope"),
```
Can we just ignore the parameter instead of panicking? I don't think a library should panic in these cases. Either we report an error or just ignore it and log a warning.

Unless the parameters are filtered somewhere else where I missed it. The input comes from the router and could make the program panic.

FloGa replied:
Ah, you are totally right, I accidently left this `panic` in the code, I intended to delete it before the PullRequest. A warning to the logger / stderr would be the proper way to go here. In case we missed arguments, we get complaints from the router anyway.

sbstp replied:
We already have the log library in the deps, `warn!` should do it.

FloGa replied:
Yep, working on it right now. ;-)

FloGa commented:
Hi Simon, I rewrote the aio.search module, this time without the Future implementation and just with `async/.await`.

sbstp commented on src/aio/gateway.rs:
```diff
@@ -200,6 +198,7 @@ impl Gateway {
         self.perform_request(
             messages::ADD_PORT_MAPPING_HEADER,
             &messages::format_add_port_mapping_message(
+                self.control_schema.get("AddPortMapping").unwrap(),
```
This unwrap is safe?

FloGa replied:
Ah, good catch. If for whatever reason the router does not report a schema for the requested action, then this call to `HashMap.get` will panic.

I rewrote it to instead propagate a dedicated RequestError variant.

sbstp commented on src/aio/gateway.rs:
```diff
@@ -245,7 +244,11 @@ impl Gateway {
                 messages::DELETE_PORT_MAPPING_HEADER,
-                &messages::format_delete_port_message(protocol, external_port),
+                &messages::format_delete_port_message(
+                    self.control_schema.get("DeletePortMapping").unwrap(),
```
Ditto?

FloGa replied:
See my [previous comment](https://github.com/sbstp/rust-igd/pull/49#discussion_r422557050). I also made the same mistake in the synchronous version of Gateway and fixed it there, too.

sbstp commented:
Looks good, I just have a few comments about the unwraps.

FloGa commented:
I rewrote the unchecked `unwrap`s and replaced them with a proper Error variant for unsupported actions. Maybe this case will never occur, but who knows, what broken hardware / firmware is out there ...

sbstp commented:
Looks! I'm gonna go a bit more testing and then merge. Are you ok with a squash merge?

FloGa commented:
> Are you ok with a squash merge?

Personally, I don't like squash merges. It creates one huge commit, that is difficult to read and understand and it makes `git bisect` nearly impossible to use, should the need arise. I won't stop you from doing one though, but in my experience you will be happier with normal merges.

sbstp commented:
I usually have a workflow that works well with squashes. I felt like I should ask here because you put a lot of care into making clean commits, so I won't squash it.

FloGa commented:
> you put a lot of care into making clean commits

Thanks, it is exactly because of the reason I mentioned and it fits well in *my* workflow.

You should choose a method that fits with *your* usual workflow, I am fine with either way. :-)

sbstp commented:
Thanks, I've published 0.11.0 with this fix.

FloGa commented:
Great to hear, thank you very much!
